fdbserver/RESTKmsConnector.actor.cpp

```diff
@@ -639,50 +777,44 @@ Future<T> kmsRequestImpl(
 
 				Reference<HTTP::IncomingResponse> resp = wait(ctx->restClient.doPost(
 				    kmsEncryptionFullUrl, requestBodyRef.toString(), RESTKmsConnectorUtils::getHTTPHeaders()));
-				curUrl->nRequests++;
+				urlCtx->nRequests++;
 
 				try {
 					T parsedResp = parseFunc(ctx, resp);
-
-					// Push urlCtx back on the ctx->urlHeap
-					while (!tempStack.empty()) {
-						ctx->kmsUrlHeap.emplace(tempStack.top());
-						tempStack.pop();
-					}
-
 					return parsedResp;
 				} catch (Error& e) {
 					TraceEvent(SevWarn, "KmsRequestRespParseFailure").error(e).detail("RequestID", requestID);
-					curUrl->nResponseParseFailures++;
+					ctx->kmsUrlStore.penalize(*urlCtx, KmsUrlCtx<KmsUrlPenaltyParams>::PenaltyType::MALFORMED_RESPONSE);
 					// attempt to do request from next KmsUrl
 				}
 			} catch (Error& e) {
-				curUrl->nFailedResponses++;
-				if (pass > 1 && isKmsNotReachable(e.code())) {
-					TraceEvent(SevWarn, "KmsRequestFailedUnreachable", ctx->uid)
+				ctx->kmsUrlStore.penalize(*urlCtx, KmsUrlCtx<KmsUrlPenaltyParams>::PenaltyType::TIMEOUT);
+				// Keep re-trying if KMS request time-out OR is server unreachable; otherwise, bubble up the error
+				if (!isKmsNotReachable(e.code())) {
+					TraceEvent(SevDebug, "KmsRequestFailedUnreachable", ctx->uid)
```

Comment:
Instead of making it SevDebug lets guard it under `FLOW_KNOBS->REST_LOG_LEVEL` and lets also add the request id to the trace

Reply:
Good idea! Will do